### DISCUSSION TITLE
RAG Agent Re-ingest During Subtask

### DISCUSSION THREAD

**Discussion by dspencej** [langroid/langroid#486] [Q&A]
I have a RAG agent that is given a directory containing documents which it ingests. The RAG agent is then asked a question from a pool of questions. It attempts to answer the question based on the information it has ingested. If it does not have enough information in its vector store to answer the question, then it passes the question to a subagent. The Subagent calls tools which generate additional documents containing more information about the question. These documents get stored in the same directory mentioned above, but never get re-ingested by the RAG agent.

My goal is to give the RAG agent a list of questions and have it use subagents to gather enough information to answer each of them. 

I would like the RAG agent to re-ingest documents after the subtask has completed.

**Comment by pchalasani** [comment]
Something very analogous to this is in the [`chat-search.py`](https://github.com/langroid/langroid/blob/main/examples/docqa/chat-search.py) example; 

There's just one agent, using two tools -- one searches the vecdb, the other does a web-search. 
The vecdb-search is attempted first, and if not useful, it does a web-search (which also ingests results into the vecdb). You should be able to modify this for your scenario (e.g. instead of web-search you'd generate additional docs relevant to the query).

**Comment by dspencej** [comment]
Thank you for the recommendation. I did look at that example initially. I'll go back and reconsider it.


Ultimately, I will have several agents each equipped with a few tools.

If the main agent doesn't have enough information to answer the question, it should select an appropriate subagent to send the question to. That agent selects a tool to run to generate a relevant document.

The process completes until all questions are answered.

My idea was that it would be better to have several subagent that specialize in a particular question area and have a smaller selection of tools, rather than having one agent with a large selection of tools.

________________________________

**Comment by pchalasani** [comment]
> If the main agent doesn't have enough information to answer the question, it should select an appropriate subagent to send the question to. That agent selects a tool to run to generate a relevant document.

I see what you mean about different agents with different specialties.

This should be easy to do with Langroid. I'll see if there are relevant examples, otherwise I may make one.
